File: solutions/typescript/protein-translation/1/protein-translation.ts
const CODON_MAP: { [key: string]: string } = {
  'AUG': 'Methionine',
  'UUU': 'Phenylalanine', 'UUC': 'Phenylalanine',
  'UUA': 'Leucine', 'UUG': 'Leucine',
  'UCU': 'Serine', 'UCC': 'Serine', 'UCA': 'Serine', 'UCG': 'Serine',
  'UAU': 'Tyrosine', 'UAC': 'Tyrosine',
  'UGU': 'Cysteine', 'UGC': 'Cysteine',
  'UGG': 'Tryptophan',
  'UAA': 'STOP', 'UAG': 'STOP', 'UGA': 'STOP'
};

export function translate(rna: string): string[] {
  const proteins: string[] = [];

  for (let i = 0; i < rna.length; i += 3) {
    const codon = rna.slice(i, i + 3);
    const protein = CODON_MAP[codon];

    if (!protein) {
      throw new Error('Invalid codon');
    }

    if (protein === 'STOP') {
      break;
    }

    proteins.push(protein);
  }

  return proteins;
}

File: solutions/kotlin/protein-translation/1/src/main/kotlin/ProteinTranslation.kt
fun translate(rna: String?): List<String> {
    if (rna == null) return emptyList()

    val proteins = mutableListOf<String>()
    
    val codons = rna.chunked(3)

    for (codon in codons) {
        val aminoAcid = when (codon) {
            "AUG" -> "Methionine"
            "UUU", "UUC" -> "Phenylalanine"
            "UUA", "UUG" -> "Leucine"
            "UCU", "UCC", "UCA", "UCG" -> "Serine"
            "UAU", "UAC" -> "Tyrosine"
            "UGU", "UGC" -> "Cysteine"
            "UGG" -> "Tryptophan"
            "UAA", "UAG", "UGA" -> "STOP"
            else -> throw IllegalArgumentException("Invalid codon")
        }

        if (aminoAcid == "STOP") break
        proteins.add(aminoAcid)
    }

    return proteins
}
